src/metabase/models/data_permissions.clj

```diff
@@ -237,6 +237,34 @@
       (or (coalesce perm-type perm-values)
           (least-permissive-value perm-type)))))
 
+(mu/defn native-download-permission-for-user :- PermissionValue
+  "Returns the effective download permission value for a given user and database ID, for native queries on the database.
+  For each group, the native download permission for a database is equal to the lowest permission level of any table in
+  the database."
+  [user-id database-id]
+  (if (is-superuser? user-id)
+    (most-permissive-value :perms/download-results)
+    (let [perm-values
+          (t2/select :model/DataPermissions
+                     {:select [[:p.perm_value :value]
+                               [:p.group_id :group_id]]
+                      :from [[:permissions_group_membership :pgm]]
+                      :join [[:permissions_group :pg] [:= :pg.id :pgm.group_id]
+                             [:data_permissions :p]   [:= :p.group_id :pg.id]]
+                      :where [:and
+                              [:= :pgm.user_id user-id]
+                              [:= :p.perm_type (u/qualified-name :perms/download-results)]
+                              [:= :p.db_id database-id]]})
+          value-by-group
+          (-> (group-by :group_id perm-values)
+              (update-vals (fn [perms]
+                             (let [values (set (map :value perms))]
```

Comment:
I added a function above called `coalesce-less-restrictive`, which I think is doing the same thing you're doing here?

Reply:
Where did you add it? Don't see it on this branch

Reply:
oh it was `coalesce-most-restrictive`. updated thanks!

Reply:
Oops, sorry I missed your earlier message. 🤦